### PR TITLE
zk3mod : added OLED

### PR DESCRIPTION
**Pull request by coomstoolbox**
## Description

Added oled variants.

## Types of Changes

- [ ] Core
- [ ] Bugfix
- [ ] New feature
- [x] Enhancement/optimization
- [ ] Keyboard (addition or update)
- [ ] Keymap/layout/userspace (addition or update)
- [ ] Documentation

## Issues Fixed or Closed by This PR

* 

## Checklist

- [x] My code follows the code style of this project: [**C**](https://docs.qmk.fm/#/coding_conventions_c), [**Python**](https://docs.qmk.fm/#/coding_conventions_python)
- [x] I have read the [**PR Checklist** document](https://docs.qmk.fm/#/pr_checklist) and have made the appropriate changes.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] I have read the [**CONTRIBUTING** document](https://docs.qmk.fm/#/contributing).
- [ ] I have added tests to cover my changes.
- [x] I have tested the changes and verified that they work and don't break anything (as well as I can manage).

### CODE REVIEW

**Review by fauxpark — CHANGES_REQUESTED** [review]
Keyboard updates must go through `develop`.

File: keyboards/handwired/ziyoulang_k3_mod/info.json [hunk 1]
```diff
@@ -11,10 +11,11 @@
         "command": false,
         "console": false,
         "extrakey": true,
-        "mousekey": true,
-        "nkro": false
+        "mousekey": false,
+        "nkro": false,
+        "lto": true
```
Comment (waffle87) [inline keyboards/handwired/ziyoulang_k3_mod/info.json:16]:
```suggestion
        "nkro": false,
        "oled": true
```

File: keyboards/handwired/ziyoulang_k3_mod/info.json [hunk 2]
```diff
@@ -11,10 +11,11 @@
...
     },
```
Comment (waffle87) [inline keyboards/handwired/ziyoulang_k3_mod/info.json:17]:
```suggestion
    },
    "build": {
        "lto": true
    },
```

File: keyboards/handwired/ziyoulang_k3_mod/keymaps/via/rules.mk [hunk 3]
```diff
@@ -1,3 +1,5 @@
 VIA_ENABLE = yes
-MOUSEKEY_ENABLE = no
-LTO_ENABLE = yes
+
+OLED_ENABLE = yes
+
+SRC += ziyoulang_k3_mod_oled.c
```
Comment (waffle87) [inline keyboards/handwired/ziyoulang_k3_mod/keymaps/via/rules.mk:5]:
```suggestion
```

File: keyboards/handwired/ziyoulang_k3_mod/readme.md [hunk 4]
```diff
@@ -23,7 +23,7 @@ See the [build environment setup](https://docs.qmk.fm/#/getting_started_build_to
 
 [KiCAD wiring diagram](https://github.com/coomstoolbox/ZK3-BP-MOD-wiring-diagram)
 
-![explanation](https://i.imgur.com/yvhvhoBh.png)
+![explanation](https://i.imgur.com/thGIdOt.png)
```
Comment (waffle87) [inline keyboards/handwired/ziyoulang_k3_mod/readme.md:26]:
```suggestion
![explanation](https://i.imgur.com/thGIdOth.png)
```

File: keyboards/handwired/ziyoulang_k3_mod/rules.mk [hunk 5]
```diff
@@ -1 +1,3 @@
-# This file intentionally left blank
+OLED_ENABLE = yes
+
+SRC += ziyoulang_k3_mod_oled.c
```
Comment (waffle87) [inline keyboards/handwired/ziyoulang_k3_mod/rules.mk:3]:
```suggestion
# This file intentionally left blank
```

File: keyboards/handwired/ziyoulang_k3_mod/ziyoulang_k3_mod_oled.c
Comment (waffle87) [inline keyboards/handwired/ziyoulang_k3_mod/ziyoulang_k3_mod_oled.c:1]: This file should be renamed to `ziyoulang_k3_mod.c`.

File: keyboards/handwired/ziyoulang_k3_mod/ziyoulang_k3_mod_oled.c [hunk 6]
```diff
@@ -0,0 +1,120 @@
+// Copyright 2023 Coom (@coomstoolbox)
+// SPDX-License-Identifier: GPL-2.0-or-later
+
+#include <stdio.h>
+
+#include QMK_KEYBOARD_H
+
```
Comment (waffle87) [inline keyboards/handwired/ziyoulang_k3_mod/ziyoulang_k3_mod_oled.c:7]:
```suggestion
```

File: keyboards/handwired/ziyoulang_k3_mod/ziyoulang_k3_mod_oled.c [hunk 7]
```diff
@@ -0,0 +1,120 @@
...
+  // Customise these values to desired behaviour
+#ifdef CONSOLE_ENABLE
+  debug_enable=true;
+  debug_matrix=true;
+#endif
+  //debug_keyboard=true;
+  //debug_mouse=true;
+
+}
```
Comment (waffle87) [inline keyboards/handwired/ziyoulang_k3_mod/ziyoulang_k3_mod_oled.c:22]:
```suggestion
void keyboard_post_init_kb(void) {
#ifdef CONSOLE_ENABLE
  debug_enable=true;
  debug_matrix=true;
#endif
  keyboard_post_init_user();
}
```

File: keyboards/handwired/ziyoulang_k3_mod/ziyoulang_k3_mod_oled.c [hunk 8]
```diff
@@ -0,0 +1,120 @@
...
+
+static uint16_t last_keycode = KC_NO;
+static uint16_t last_col = 0;
+static uint16_t last_row = 0;
+
+bool process_record_user(uint16_t keycode, keyrecord_t *record) {
+   // コンソールが有効化されている場合、マトリックス上の位置とキー押下状態を出力します
```
Comment (waffle87) [inline keyboards/handwired/ziyoulang_k3_mod/ziyoulang_k3_mod_oled.c:29]:
```suggestion
bool process_record_kb(uint16_t keycode, keyrecord_t *record) {
    if (!process_record_user(keycode, record)) {
        return false;
    }
```

File: keyboards/handwired/ziyoulang_k3_mod/ziyoulang_k3_mod_oled.c [hunk 9]
```diff
@@ -0,0 +1,120 @@
...
+        if (last_keycode != keycode) {
+            last_keycode = keycode;
+            last_col = record->event.key.col;
+            last_row = record->event.key.row;
+        }
+    }
+    return true;
+}
+
+oled_rotation_t oled_init_user(oled_rotation_t rotation) {
```
Comment (waffle87) [inline keyboards/handwired/ziyoulang_k3_mod/ziyoulang_k3_mod_oled.c:43]:
```suggestion
oled_rotation_t oled_init_kb(oled_rotation_t rotation) {
```

File: keyboards/handwired/ziyoulang_k3_mod/ziyoulang_k3_mod_oled.c [hunk 10]
```diff
@@ -0,0 +1,120 @@
...
+    };
+    oled_write_P(qmk_logo, false);
+#ifdef CONSOLE_ENABLE
+    uprintf("Ziyoulang K3 Mod\n");
+    oled_set_cursor(3, 0);
+    oled_write_P(PSTR("Ziyoulang K3 Mod"), false);
+#endif
+}
+
+bool oled_task_user(void) {
```
Comment (waffle87) [inline keyboards/handwired/ziyoulang_k3_mod/ziyoulang_k3_mod_oled.c:61]:
```suggestion
bool oled_task_kb(void) {
    if (!oled_task_user()) {
        return false;
    }
```

**Review by waffle87 — CHANGES_REQUESTED** [review]

**Comment by coomstoolbox** [comment]
@waffle87 
I appreciate the many confirmations. 
I understand these and will fix them.

File: keyboards/handwired/ziyoulang_k3_mod/ziyoulang_k3_mod.h [hunk 11]
```diff
@@ -0,0 +1,6 @@
+// Copyright 2023 Coom (@coomstoolbox)
```
Comment (fauxpark) [inline keyboards/handwired/ziyoulang_k3_mod/ziyoulang_k3_mod.h:1]: This file can be deleted.

File: keyboards/handwired/ziyoulang_k3_mod/ziyoulang_k3_mod.c [hunk 12]
```diff
@@ -0,0 +1,122 @@
+// Copyright 2023 Coom (@coomstoolbox)
+// SPDX-License-Identifier: GPL-2.0-or-later
+
+#include "ziyoulang_k3_mod.h"
```
Comment (fauxpark) [inline keyboards/handwired/ziyoulang_k3_mod/ziyoulang_k3_mod.c:4]:
```suggestion
#include "quantum.h"
```

File: keyboards/handwired/ziyoulang_k3_mod/ziyoulang_k3_mod.c [hunk 13]
```diff
@@ -0,0 +1,122 @@
...
+#include <stdio.h>
+#ifdef CONSOLE_ENABLE
+   #include <print.h>
+#endif
```
Comment (fauxpark) [inline keyboards/handwired/ziyoulang_k3_mod/ziyoulang_k3_mod.c:8]:
```suggestion
```

**Review by fauxpark — CHANGES_REQUESTED** [review]

**Review by waffle87 — APPROVED** [review]

File: keyboards/handwired/ziyoulang_k3_mod/ziyoulang_k3_mod.c [hunk 14]
```diff
@@ -0,0 +1,119 @@
...
+#ifdef CONSOLE_ENABLE
+    debug_enable=true;
+    debug_matrix=true;
+#endif
+    keyboard_post_init_user();
+}
+
+static uint16_t last_keycode = KC_NO;
+static uint16_t last_col = 0;
+static uint16_t last_row = 0;
```
Comment (drashna) [inline keyboards/handwired/ziyoulang_k3_mod/ziyoulang_k3_mod.c:17]: both of these are uint8_t, and may be simpler to use that.  Also, you could use keypos_t, as that's just a structure with both col and row.

File: keyboards/handwired/ziyoulang_k3_mod/ziyoulang_k3_mod.c [hunk 15]
```diff
@@ -0,0 +1,119 @@
...
+    oled_write_P(PSTR("kc : "), false);
+    if (last_keycode > 21000) {
+      oled_write_ln_P(PSTR("    Fn"), false);
+    } else {
+      oled_write_ln_P(get_u16_str(last_keycode, ' '), false);
+    }
+    oled_write_P(PSTR("col: "), false);
+    oled_write_P(get_u16_str(last_col, ' '), false);
+    oled_write_P(PSTR(",row: "), false);
+    oled_write_P(get_u16_str(last_row, ' '), false);
```
Comment (drashna) [inline keyboards/handwired/ziyoulang_k3_mod/ziyoulang_k3_mod.c:115]: You can/should use `get_u8_str` here, after changing the variables as these are only over uint8_t, internally.

**Comment by coomstoolbox** [comment]
Is there any new code to correct? I would like to proceed to the next step if there are no problems.

**Pull request merged**